wagtail/contrib/search_promotions/tests.py

```diff
@@ -280,6 +345,75 @@ def test_post_without_recommendations(self):
             "Please specify at least one recommendation for this search term.",
         )
 
+    def test_post_with_page_and_external_link(self):
+        post_data = {
+            "query_string": "test",
+            "editors_picks-TOTAL_FORMS": 1,
+            "editors_picks-INITIAL_FORMS": 0,
+            "editors_picks-MAX_NUM_FORMS": 1000,
+            "editors_picks-0-DELETE": "",
+            "editors_picks-0-ORDER": 0,
+            "editors_picks-0-page": 1,
+            "editors_picks-0-external_link_url": "https://wagtail.org",
+            "editors_picks-0-external_link_text": "Wagtail",
+            "editors_picks-0-description": "Hello",
+        }
+        response = self.client.post(reverse("wagtailsearchpromotions:add"), post_data)
+
+        # User should be given an error
+        self.assertEqual(response.status_code, 200)
+        self.assertFormsetError(
+            response,
+            "searchpicks_formset",
+            None,
+            None,
+            "Please only select a page OR enter an external link.",
```

Comment:
Thank you for pulling these out from variable usage to 'hard-coding' in the unit tests.

While this can seem like duplication I just wanted to explain my reasoning for pushing for this approach instead of referencing a property on the class (as originally proposed in the PR). Others may disagree but I wanted to explain the 'why' so you see where I we were coming from.

1. Accidental errors, if you used a property such as `MyModel.MY_CONSTANT` and it is somehow malformed or wrong (e.g. an invalid string or not using translations nicely) and then your test says 'check error equals `MyModel.MY_CONSTANT`'. This can hide legitimate errors, forcing the test to type out the desired value makes the test find these kinds of errors easier.
2. Unit tests that do not know about 'implementation'. While the boundary can be blurry, it's nice to be able to test against the desired output and ignore implementation. By testing against our ouput error, the code that generates/stores/contains this error can be changed without changing the unit test.
3. Leaves room for future variable based translations, imagine we wanted to make this error based on some data (e.g. 'please select a page OR an external link, you have selected page 5'). Making this change requires our tests to look at partially re-implementing logic of this data usage OR just change to checking the output string anyway.

While it's not a hard rule, as in I am sure you will find tests that do 1:1 comparison with values for things like this, we tend to test against the output result and try to avoid our tests knowing too much about the implementation. This does lead to some duplication, but in my experience it means that the tests uncover more legitimate errors and also the tests only need to change when the output changes, not when the implementation driving the output changes.

I hope it's OK to give this feedback, I just wanted to make sure all your points of concern were addressed where possibile.